server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,300 @@
+            )
+        );
+
+        static {
+            SPLIT_PARSER.declareObjectArray(
+                ConstructingObjectParser.constructorArg(),
+                (parser, c) -> SourceShardState.valueOf(parser.text()),
+                SOURCE_SHARDS_FIELD
+            );
+            SPLIT_PARSER.declareObjectArray(
+                ConstructingObjectParser.constructorArg(),
+                (parser, c) -> TargetShardState.valueOf(parser.text()),
+                TARGET_SHARDS_FIELD
+            );
+        }
+
+        private final int oldShardCount;
+        private final int newShardCount;
+        private final SourceShardState[] sourceShards;
+        private final TargetShardState[] targetShards;
+
+        Split(SourceShardState[] sourceShards, TargetShardState[] targetShards) {
+            this.sourceShards = sourceShards;
+            this.targetShards = targetShards;
+
+            oldShardCount = sourceShards.length;
+            newShardCount = oldShardCount + targetShards.length;
+        }
+
+        Split(StreamInput in) throws IOException {
+            this(
+                in.readArray(i -> i.readEnum(SourceShardState.class), SourceShardState[]::new),
+                in.readArray(i -> i.readEnum(TargetShardState.class), TargetShardState[]::new)
+            );
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeArray(sourceShards);
+            out.writeArray(targetShards);
+        }
+
+        static Split fromXContent(XContentParser parser) throws IOException {
+            return SPLIT_PARSER.parse(parser, null);
+        }
+
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(SOURCE_SHARDS_FIELD.getPreferredName(), sourceShards);
+            builder.field(TARGET_SHARDS_FIELD.getPreferredName(), targetShards);
+
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+            Split otherState = (Split) other;
+            // we can ignore oldShardCount and newShardCount since they are derived
+            return Arrays.equals(sourceShards, otherState.sourceShards) && Arrays.equals(targetShards, otherState.targetShards);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(Arrays.hashCode(sourceShards), Arrays.hashCode(targetShards));
+        }
+
+        public int oldShardCount() {
+            return oldShardCount;
+        }
+
+        public int newShardCount() {
+            return newShardCount;
+        }
+
+        public SourceShardState[] sourceShards() {
+            return sourceShards;
+        }
+
+        public TargetShardState[] targetShards() {
```

Comment:
It is not entirely clear to me if this is indexed by shardId or by (shardId-sourceShards)? I.e., we go from 2 to 4 shards, we have source shards 0-1 and target shards 2-3, what is the return value here?

I wonder if we need to expose it as arrays? Perhaps it is better to do `targetShard(shardId)` in the client code?

Reply:
I see it now in the method below. I think exposing it as an array is somewhat involved to callers?

Reply:
I also see the getter, so I propose to just remove these or make them more private.

Reply:
yes, thanks, these are there just for the serialization tests really. I forgot to adjust their access modifiers, will do that now.

Reply:
Done in 0e69a76